src/compiler/checker.ts

```diff
@@ -23029,6 +23029,18 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             type;
     }
 
+    function getValueOfResult(type: Type): Type {
+        const valueOf = getPropertyOfType(type, "valueOf" as __String);
+        if (valueOf) {
+            const signatures = getSignaturesOfType(getTypeOfSymbol(valueOf), ts.SignatureKind.Call);
+            if (signatures && signatures.length > 0) {
+                const returnType = getReturnTypeOfSignature(signatures[0]);

```

Comment:
This favors the first signature in an overload list, while usually in places where we're overload-blind today we usually favor the last overload in the list. Maybe `signatures[signatures.length - 1]` for consistency? Or, potentially better yet, rather than picking one overload, `getUnionType(map(signatures, getReturnTypeOfSignature))` just to hedge (in theory, only overloads that match an all-undefined/empty argument list would actually apply.... but overloads on `valueOf` are probably a bit suspicious anyway). I say that, but we'll actually see an overloaded `valueOf` for every intersected type eg, `number & {myBrand}` is going to have a `valueOf` from `number`'s apparent type and `object`'s apparent type intersected together, which'll result in two overloads unless they're merged for being identical.